### ISSUE TITLE
Add generic Type-C "port manager" support

### ISSUE THREAD

**Issue by raveslave** [zephyrproject-rtos/zephyr#18579]
**Add support for USB Type-C controllers with PD**

This is useful when designing products using USB-C connectors needing additional support like higher current, role-switching "host/device", power source/sink or simply supporting vendor packages over PD (using the CC pins).

The driver should implement support for
- DRP (DFP/UFP), i.e. the type-c terms for "OTG", with low amount of programmability.
- PD 2.0 / 3.0 to support source or sink, incl providing or negotiating power-requirements (voltage/current) with PD-compliant power-supplies or computers using Type-C.
- Alternate mode(s)

The driver should basically serve as a state machine that orchestrates connection-events, data and power roles as well as runtime changes like role-swap or power-updates.

The connection between the TCPM (running in Zephyr) and the TCPC (vendor IC) is defined as the USB Type-C Port Controller Interface, TCPCI. 
see: https://www.usb.org/document-library/usb-type-cr-port-controller-interface-specification

**Event and control needs**

The low level driver should be responsible for
- Reporting VBUS status and activating VBUS
- Setting CC lines and providing CC line status
- Setting line polarity
- Activating and deactivating VCONN
- Setting current limits
- Activating and deactivating PD message transfers
- Sending and receiving PD messages
- wip...

---   
<img width="550" alt="image" src="https://user-images.githubusercontent.com/1734770/63512791-e5e3a600-c4e4-11e9-83ea-e15a8fe73eb1.png">

**Comment by keith-zephyr** [comment]
Google is in the planning process of upstreaming the [USB-C stack](https://source.chromium.org/chromiumos/chromiumos/codesearch/+/main:src/platform/ec/common/usbc/) used on Chromebooks into Zephyr.

The is a bit of work needed to reorganize the code to conform to Zephyr's structure, so if anyone has specific features like would like prioritized, please add a comment to this issue.

**Comment by dzarda** [comment]
@keith-zephyr That's a nice bit of news. Last time I checked there was no Sink-only support in the Chrome OS codebase. However that's a major case for Zephyr devices. Would that be available?

**Comment by keith-zephyr** [comment]
Yes, we plan to provide source-only, sink-only, and dual-role port operation.  We also support try source and try sink operation for dual-role ports.

**Comment by keith-zephyr** [comment]
This [github issue](https://github.com/zephyrproject-rtos/zephyr/issues/38371) has been created that will cover the work of moving Google's USB-C stack to Zephyr.   We welcome comments and suggestions.

**Comment by gmsdesai87** [comment]
I’m trying to understand the control boundary between TCPM (policy engine) and the TCPC/PMIC in real systems (especially Android/Linux phones).

From this discussion, it looks like TCPM handles policy decisions like role selection and PD negotiation, while the TCPC handles signaling, CC state, and safety enforcement.

In practice, during something like a USB-C phone-to-phone connection:

- If the OS/TCPM requests a power role swap, can the TCPC/PMIC refuse or override it?
- Are there cases where hardware constraints (thermal, current limits, protection logic) take final authority regardless of the policy engine’s request?

Would love to understand how “final authority” is actually split between software (TCPM) and hardware (controller/PMIC) in real implementations.

**Comment by gmsdesai87** [comment]
When you say the driver acts as a state machine orchestrating roles — does that mean TCPM has full control over role transitions, or can the underlying TCPC/PMIC still block or modify those transitions based on hardware conditions?

**Comment by keith-zephyr** [comment]
> I’m trying to understand the control boundary between TCPM (policy engine) and the TCPC/PMIC in real systems (especially Android/Linux phones).
> 
> From this discussion, it looks like TCPM handles policy decisions like role selection and PD negotiation, while the TCPC handles signaling, CC state, and safety enforcement.

This is basically  correct.  The TCPC only handles the CC state.  Any PD messages received by the TCPC are passed on to the TCPM for processing.   If the TCPM needs to send a PD message to the port partner, it uses `tcpc_transmit_data()` to send the message.

The TCPM handles policy decisions, but your application can also control policy through the extensive set of callbacks in `include/zephyr/usb_c/usbc.h`


> 
> In practice, during something like a USB-C phone-to-phone connection:
> 
> * If the OS/TCPM requests a power role swap, can the TCPC/PMIC refuse or override it?

No, the TCPC should not make policy decisions.   In the case where  a power role swap is initiated by the application, the swap request is sent to the connected partner device to accept/reject.  When the partner device sends a power role swap request to us, the TCPC doesn't so any processing of the incoming message.  It informs the TCPM that a new receive message is available and the TCPM reads the message using `tcpc_get_rx_pending_msg()` and does all the processing, including making the decision whether to accept or reject the power role swap.

> * Are there cases where hardware constraints (thermal, current limits, protection logic) take final authority regardless of the policy engine’s request?

No - any hardware constraints for your board should be managed using the policy callback noted above.

> 
> Would love to understand how “final authority” is actually split between software (TCPM) and hardware (controller/PMIC) in real implementations.

See above, the final authority is your application code.

**Comment by keith-zephyr** [comment]
> When you say the driver acts as a state machine orchestrating roles — does that mean TCPM has full control over role transitions, or can the underlying TCPC/PMIC still block or modify those transitions based on hardware conditions?

The TCPM, in conjunction with your application's policy callbacks controls the permitted transitions.  The TCPC is really just a PHY layer for sending/receiving PD messages.

**Comment by gmsdesai87** [comment]
Thanks, this really helped clarify the roles of TCPM, TCPC, and PMIC.

I wanted to go one level deeper on the design side:

If TCPM is primarily handling protocol execution and reacting to events, is it correct to think that higher-level “policy decisions” (like when to initiate a role swap or limit power transfer based on system conditions) are expected to be implemented outside TCPM?

In real systems (like Android/Linux-based devices), where does this policy layer typically live?

For example:
- Is it handled in kernel space alongside TCPM?
- Or in a higher-level system service / userspace layer?
- Or is it vendor-specific?

I’m trying to understand the intended separation between:
1. protocol execution (TCPM)
2. hardware enforcement (PMIC/TCPC)
3. decision-making (policy layer)

Appreciate any guidance or pointers.

**Comment by keith-zephyr** [comment]
> Thanks, this really helped clarify the roles of TCPM, TCPC, and PMIC.
> 
> I wanted to go one level deeper on the design side:
> 
> If TCPM is primarily handling protocol execution and reacting to events, is it correct to think that higher-level “policy decisions” (like when to initiate a role swap or limit power transfer based on system conditions) are expected to be implemented outside TCPM?
> 

This is correct, policy decisions are expected to be managed by the application.

> In real systems (like Android/Linux-based devices), where does this policy layer typically live?
> 
> For example:
> 
> * Is it handled in kernel space alongside TCPM?
> * Or in a higher-level system service / userspace layer?
> * Or is it vendor-specific?

Ideally, the policy callbacks are implemented in userspace, but I'm not sure if this has been tested.

I recommend looking at the [USB-C samples](https://github.com/zephyrproject-rtos/zephyr/tree/9be6ee7d642cc3c1448e2c0500b0720a1748d3ce/samples/subsys/usb_c) for how this is all connected.

There are 3 samples: source-only application, sink-only, and a DRP (dual-role-power).

**Comment by gmsdesai87** [comment]
Thanks, this clarifies the TCPM/TCPC boundary. I have one follow-up question related to a USB-C phone-to-phone power-sharing use case.
Suppose an application wants to implement a user-controlled "Send Battery" function where:
Device A selects a target amount of energy to transfer to Device B;
Device B accepts the request;
the system establishes the appropriate source/sink and PD configuration;
the transfer is monitored; and
the transfer terminates when the agreed energy target has been reached.
In the Zephyr/Linux architecture, which layer would normally be responsible for:
receiving the application's requested energy-transfer target;
translating that target into the appropriate PD/power policy;
measuring or determining how much energy has actually been delivered; and
terminating the transfer when the target is reached?
In particular, is there an existing standard/API mechanism for an application to request a quantified energy-transfer target (for example Wh or an equivalent battery-energy target), or does the application normally only request/affect power roles and electrical operating parameters?
I'm trying to understand whether a user-level application can realistically sit above TCPM and control a quantified peer-to-peer energy-transfer transaction, or whether that requires OEM/platform-specific integration.

**Comment by keith-zephyr** [comment]
> In particular, is there an existing standard/API mechanism for an application to request a quantified energy-transfer target (for example Wh or an equivalent battery-energy target), or does the application normally only request/affect power roles and electrical operating parameters? I'm trying to understand whether a user-level application can realistically sit above TCPM and control a quantified peer-to-peer energy-transfer transaction, or whether that requires OEM/platform-specific integration.

I'm not aware of anything in the PD specification that tracks or communicates information about cumulative energy.  So this kind of policy needs to be tracked and managed by the application.

If the application is monitoring the energy transfer (or more likely just monitoring the current battery charge), the sink device can stop accepting current by disabling the `SinkVbus` in the TCPC.  The PD contract remains in place so VBUS remains at the negotiated voltage, but the sink device just doesn't accept current.  The Zephyr TCPC API `tcpc_set_snk_ctrl()` can be used to enable and disable sinking.

For systems that have multiple USB-C ports (such as a laptop), it's common that you disable VBUS sinking as a default.  Only after evaluating the PD contracts on all ports (if connected), the application then selects the "best" port to charge from and only enables VBUS sinking on the specific port.  Enabling VBUS sinking on multiple ports is likely to cause damage to the device.
